### PR TITLE
Promote experimental proxy-status command to regular, and remove regular `proxy-status` command

### PR DESCRIPTION
**Pull request by hanxiaop**
**Please provide a description of this PR:**
Plan to ~remove~ promote the `experimental proxy-status` command in the next release. This command mirrors the `proxy-status` command, necessitating duplicate efforts in maintenance whenever the `proxy-status`  command is updated.

There were some discussions related to this: https://github.com/istio/istio/pull/46050

### CODE REVIEW

**Review by hzxuzhonghu — APPROVED** [review]

File: tests/integration/pilot/piggyback_test.go
Comment (GregHanson) [inline tests/integration/pilot/piggyback_test.go:78]: the other integration test was a duplicate of the base `istioctl ps`, but are we loosing coverage in this one by removing this test?  is it no longer valid or is it covered somewhere else?

Comment (GregHanson) [reply tests/integration/pilot/piggyback_test.go:78]: I think some of this was discussed here: https://github.com/istio/istio/pull/46050#issuecomment-1661953633

if this is a loss of functionality - do we want to remove the test, or do we want to create an issue for tracking and use a `t.Skip(<link to GitHub issue>)` ?

Comment (hanxiaop) [reply tests/integration/pilot/piggyback_test.go:78]: You are right, this will lead to less coverage I think. I think before doing `t.Skip` or creating an issue, I'll look into the code, and try to replace it with a new test method to test the gPRC-based Tap instead of `istioctl x ps`.

Comment (hanxiaop) [reply tests/integration/pilot/piggyback_test.go:78]: This should be addressed by https://github.com/istio/istio/pull/47801.

**Review by GregHanson — DISMISSED** [review]

**Comment by howardjohn** [comment]
I want to do some more investigating usage before merging. will report back soon

**Comment by howardjohn** [comment]
also the point of the experimental command was to replace the old one. This does the opposite. What changed?

**Comment by hanxiaop** [comment]
> also the point of the experimental command was to replace the old one. This does the opposite. What changed?

@howardjohn The intention is to avoid dup efforts in maintaining two commands `ps` and `experimental ps` since they are quite similar at this moment. If there are any concerns about removing the experimental command, we can also replace it with the experimental command, which provides more ways to get the proxy status. I'm looking forward to see the usage.

**Comment by ingwonsong** [comment]
> also the point of the experimental command was to replace the old one.

big +1 to promoting the experimental one to regular, if it is necessary.

it is because:

1. There would be the reason why we had to implement the experimental one even though it was "re-inventing the wheel"
2. The experimental one provides various way to access the `proxy-status` information, which means much enhanced than the old one.

**Comment by hanxiaop** [comment]
> 1. There would be the reason why we had to implement the experimental one even though it was "re-inventing the wheel"
> 2. The experimental one provides various way to access the `proxy-status` information, which means much enhanced than the old one.

IIRC the original intention was to replace the original `ps` command. Since `x ps` is also well-maintained, I think promoting it is appropriate. Currently, `ps` should be considered a subset of `x ps`.

**Comment by hanxiaop** [comment]
@GregHanson @howardjohn @ingwonsong Can any of you bring this to the WG meeting to gather consensus on the promotion? Thanks.

**Comment by GregHanson** [comment]
@hanxiaop added to the agenda for next week's community meeting - since this week's meeting is cancelled due to US holiday this week

**Comment by hanxiaop** [comment]
> @hanxiaop added to the agenda for next week's community meeting - since this week's meeting is cancelled due to US holiday this week

Sounds good, thanks

**Comment by GregHanson** [comment]
discussion from 11/29 community call: ([notes](https://docs.google.com/document/d/1wsa06GGiq1LEGwhkiPP0FKIZJqdAiue-VeBonWAzAyk/edit?pli=1#heading=h.3lzzxgz0kjhk))

TLDR;
 - general approval for the promotion

TODO's
 - verify test coverage of `x ps`
 - `x ps` needs to remain for N releases before finally being removed
 - long term follow-up: open issue for tracking deprecation of the HTTP debug API on istiod
   - for backwards compatibility, older istioctl binaries can still be used to access the original `ps` command

File: releasenotes/notes/47740.yaml
Comment (GregHanson) [inline releasenotes/notes/47740.yaml:6]: Should probably clarify the release note a little more.  Add something along the lines of `This promotion should not result in any loss of functionality.  Adds additional flags to support multi-control plane environments`

**Comment by hanxiaop** [comment]
@GregHanson I think this PR meets most of the TODO's, and I'll create an issue to track the HTTP debug API.

> verify test coverage of x ps

I think that currently, both `ps` and `x ps` are being tested by this PR. Although they are the same, we need to maintain `x ps` for several releases, so I prefer testing both which makes more sense. See the integration test changes for examples.

One question is: do we need to add tests for the xDS related flags? I couldn't find any existing tests for these options, which were introduced in https://github.com/istio/istio/pull/25087 without any tests. BTW there are already graduated commands like `istioctl kube-inject` that use these flags. Besides the flags, I think the command is covered with tests.

**Review by GregHanson — APPROVED** [review]
looks good to me, but I will wait for any last comments from @howardjohn before removing the `no-not-merge` label

File: releasenotes/notes/47740.yaml
Comment (linsun) [inline releasenotes/notes/47740.yaml:8]: Could you explain the existing `istioctl ps` cmd?  To readers, it is confusing we have promoted the `istioctl x ps` while we already has `istioctl ps` today.

**Review by linsun — COMMENTED** [review]
RN needs some work

**Comment by howardjohn** [comment]
`x ps` has ~3 modes:

1. Query each istiod and portforward
2. xds-via-agents
3. xds-address

(1) just changes from HTTP to grpc/xds, which is not a concern and seems like a good approach. 2/3 I am less sure of are at "stable" yet (especially 2).

Do we want to promote all 3?

Hypothetically we could keep 1 implementation as `ps` and `x ps`, but disallow some flags under the stable `ps`

**Comment by hanxiaop** [comment]
> `x ps` has ~3 modes:
> 
> 1. Query each istiod and portforward
> 2. xds-via-agents
> 3. xds-address
> 
> (1) just changes from HTTP to grpc/xds, which is not a concern and seems like a good approach. 2/3 I am less sure of are at "stable" yet (especially 2).
> 
> Do we want to promote all 3?
>
> Hypothetically we could keep 1 implementation as `ps` and `x ps`, but disallow some flags under the stable `ps`

1 should not be a problem. I think 3 won't be either, as it is quite similar to 1 but applies directly to the xds-address. The xds-address has been tested and is already used in graduated commands like `kube-inject`. I agree there may be uncertainty about 2 and seems that disallowing some flags would be a great approach to gradually promote the command.

I prefer to hold off on this until 1.22 to allow more time for research and testing, which will help us gain confidence.

File: releasenotes/notes/47740.yaml
Comment (hanxiaop) [reply releasenotes/notes/47740.yaml:8]: Great suggestion! I'll revise it.

**Comment by hanxiaop** [comment]
not stale

**Comment by hanxiaop** [comment]
> `x ps` has ~3 modes:
> 
> 1. Query each istiod and portforward
> 2. xds-via-agents
> 3. xds-address
> 
> (1) just changes from HTTP to grpc/xds, which is not a concern and seems like a good approach. 2/3 I am less sure of are at "stable" yet (especially 2).
> 
> Do we want to promote all 3?
> 
> Hypothetically we could keep 1 implementation as `ps` and `x ps`, but disallow some flags under the stable `ps`

@howardjohn I've marked option 2 as hidden in https://github.com/istio/istio/pull/47740/commits/7e1366c553f917c7b714c9c70ae143e9ef7847ed. For option 3, the flags already exist in graduated flags like `kube-inject`. Although the logic is not identical, I have tested it and verified that it can work as expected. I think if there are any bugs, we can do a forward fix. Let me know if this addresses your concerns.

File: istioctl/pkg/proxystatus/proxystatus_test.go
Comment (howardjohn) [inline istioctl/pkg/proxystatus/proxystatus_test.go:50]: seems we are removing valid test coverage here?

**Review by howardjohn — APPROVED** [review]
LGTM just one test comment. We should merge https://github.com/istio/istio/pull/49728 too

File: istioctl/pkg/proxystatus/proxystatus_test.go
Comment (hanxiaop) [reply istioctl/pkg/proxystatus/proxystatus_test.go:50]: The test is incorrect because the fake client returns an empty response with no error. In a real cluster, errors are shown for both commands when there is no Istiod pod running.
```
➜  istio git:(master)  istioctl ps                                 
Error: unable to find any Istiod instances
➜  istio git:(master) istioctl x ps
Error: no running Istio pods in "istio-system"
```

Comment (howardjohn) [reply istioctl/pkg/proxystatus/proxystatus_test.go:50]: every test case now gives errors, can we have a test that covers success?

Comment (hanxiaop) [reply istioctl/pkg/proxystatus/proxystatus_test.go]: Added, thanks!

**Pull request merged**
